docs/good-fuzz-target.md

```diff
@@ -0,0 +1,244 @@
+* It is being executed as part of the usual testing process using the [seed corpus](#seed-corpus) as inputs.
+* Inputs that trigger bugs on the fuzz target are added to the [seed corpus](#seed-corpus) as a form of regression testing.
+
+
+# One-time initialization
+Sometimes an API under test needs to be initialized.
+
+The simplest and the recommended way is to have a statically initialized global object inside LLVMFuzzerTestOneInput:
+
+```cpp
+// fuzz_target.cc
+extern "C" int LLVMFuzzerTestOneInput(const uint8_t *Data, size_t Size) {
+  static bool Initialized = DoInitialization();
+  ...
+```
+
+Or (less preferable) in the global scope if that works for you:
+
+```cpp
+// fuzz_target.cc
+static bool Initialized = DoInitialization();
+extern "C" int LLVMFuzzerTestOneInput(const uint8_t *Data, size_t Size) {
+  ...
+```
+
+# Determinism
+
+A fuzz target needs to be deterministic, i.e. given the same input it should
+have the same behavior.
+This means, for example, that a fuzz target should not use `rand()` or any other
+source of randomness.
+
+
+# Speed
+
+Fuzzing is a search algorithm that requires **many** iterations, and so
+a good fuzz target should be very fast.
+
+A typical good fuzz target will have an order of 1000 executions per second per
+one CPU core on average (exec/s) or more. For lightweight targets, 10000 exec/s or more.
+
+If your fuzz target has less than 10 exec/s you are probably doing something
+wrong. We recommend to profile fuzz targets and eliminate any obvious hot
+spots.
+
+# Memory consumption
+
+For CPU-efficient fuzzing a good fuzz target should consume less RAM than it is
+available on a given (virtual) machine per one CPU core.
+There is no one-size-fits-all RAM threshold, but as of 2019 a typical good fuzz
+target would consume less than 1.5Gb.
+
+# Global state
+
+Ideally, a fuzz target should not modify a global state because otherwise
+the fuzz target may be non-deterministic.
+
+It may not be possible in every case
+(strictly speaking, even calling `malloc()` modifies a global state).
+
+# Timeouts, OOMs, shallo bugs
+
+A good fuzz target should not have any
+timeouts (inputs that take too long to process),
+OOMs(input that cause the fuzz target to consume too much RAM),
+or shallow (easily discoverable) bugs.
+Otherwise fuzzing will stall quickly.
+
+# Seed corpus
+
+In most cases a good fuzz target should be accompanied with a *seed corpus*,
+which is a set of representative inputs for the fuzz target.
+These inputs combined should cover large portions of the API under test,
+ideally achieving 100% coverage (different coverage metrics can be applied, e.g.
+block coverage or edge coverage, depending on a specific case).
+
+Avoid large seed inputs when smaller inputs are sufficient for providing the
+same coverage.
+
+A seed corpus is stored as a directory where every individual file represents one input.
```

Comment:
nit: maybe add "The inputs can be put in subdirectories inside the main corpus dir." ?